File: FacturXDotNet/InvoiceTypeCode.cs
namespace FacturXDotNet;

/// <summary>
///     <b>Invoice type code</b> - A code specifying the functional type of the Invoice.
/// </summary>
/// <remarks>
///     Commercial invoices and credit notes are defined according the entries in UNTDID 1001. Other  entries of UNTDID 1001 with specific invoices or credit notes may be used if
///     applicable.
/// </remarks>
/// <ID>BT-3</ID>
/// <BR-4>An Invoice shall have an Invoice type code.</BR-4>
/// <CiiXPath>/rsm:CrossIndustryInvoice/rsm:ExchangedDocument/ram:TypeCode</CiiXPath>
/// <Profile>MINIMUM</Profile>
public enum InvoiceTypeCode
{
    CommercialInvoice = 380,
    CreditNote = 381,
    CorrectionInvoice = 384,
    SelfBilledInvoice = 389,

    /// <remarks>
    ///     Not accepted by Chorus Pro.
    /// </remarks>
    SelfBilledCreditNote = 261,
    PrepaymentInvoice = 386,

    /// <remarks>
    ///     Not accepted by Chorus Pro.
    /// </remarks>
    InvoiceInformationForAccountingPurpose = 751
}

public static class FacturXTypeCodeMappingExtensions
{
    public static int ToSpecificationIdentifierString(this InvoiceTypeCode value) =>
        value switch
        {
            InvoiceTypeCode.CommercialInvoice => 380,
            InvoiceTypeCode.CreditNote => 381,
            InvoiceTypeCode.CorrectionInvoice => 384,
            InvoiceTypeCode.SelfBilledInvoice => 389,
            InvoiceTypeCode.SelfBilledCreditNote => 261,
            InvoiceTypeCode.PrepaymentInvoice => 386,
            InvoiceTypeCode.InvoiceInformationForAccountingPurpose => 751,
            _ => throw new ArgumentOutOfRangeException(nameof(value), value, null)
        };

    public static InvoiceTypeCode? ToSpecificationIdentifierOrNull(this int value) =>
        value switch
        {
            380 => InvoiceTypeCode.CommercialInvoice,
            381 => InvoiceTypeCode.CreditNote,
            384 => InvoiceTypeCode.CorrectionInvoice,
            389 => InvoiceTypeCode.SelfBilledInvoice,
            261 => InvoiceTypeCode.SelfBilledCreditNote,
            386 => InvoiceTypeCode.PrepaymentInvoice,
            751 => InvoiceTypeCode.InvoiceInformationForAccountingPurpose,
            _ => null
        };

    public static InvoiceTypeCode ToSpecificationIdentifier(this int value) =>
        value.ToSpecificationIdentifierOrNull() ?? throw new ArgumentOutOfRangeException(nameof(InvoiceTypeCode), value, null);
}

File: src/FacturXDotNet.WebEditor/src/app/core/facturx-models/cii/invoice-type-code.ts
/**
 * **Invoice type code** - A code specifying the functional type of the Invoice.
 *
 * @remarks
 * Commercial invoices and credit notes are defined according the entries in UNTDID 1001. Other  entries of UNTDID 1001 with specific invoices or credit notes may be used if
 * applicable.
 *
 * @ID BT-3
 * @BusinessRules **BR-4**: An Invoice shall have an Invoice type code.
 * @CiiXmlPath /rsm:CrossIndustryInvoice/rsm:ExchangedDocument/ram:TypeCode
 * @Profile MINIMUM
 * @ChorusPro The types of documents used are:
 *     - 380: Commercial Invoice
 *     - 381: Credit note
 *     - 384: Corrected invoice
 *     - 389: Self-billed invoice (created by the buyer on behalf of the supplier)
 *     - 261: Self billed credit note (not accepted by CHORUSPRO)
 *     - 386: Prepayment invoice
 *     - 751: Invoice information for accounting purposes (not accepted by CHORUSPRO)
 */
export type InvoiceTypeCode =
  | '71-request-for-payment'
  | '80-debit-note-related-to-goods-or-services'
  | '81-credit-note-related-to-goods-or-services'
  | '82-metered-services-invoice'
  | '83-credit-note-related-to-financial-adjustments'
  | '84-debit-note-related-to-financial-adjustments'
  | '102-tax-notification'
  | '130-invoicing-data-sheet'
  | '202-direct-payment-valuation'
  | '203-provisional-payment-valuation'
  | '204-payment-valuation'
  | '211-interim-application-for-payment'
  | '218-final-payment-request-based-on-completion-of-work'
  | '219-payment-request-for-completed-units'
  | '261-self-billed-credit-note'
  | '262-consolidated-credit-note---goods-and-services'
  | '295-price-variation-invoice'
  | '296-credit-note-for-price-variation'
  | '308-delcredere-credit-note'
  | '325-proforma-invoice'
  | '326-partial-invoice'
  | '331-commercial-invoice-which-includes-a-packing-list'
  | '380-commercial-invoice'
  | '381-credit-note'
  | '382-commission-note'
  | '383-debit-note'
  | '384-corrected-invoice'
  | '385-consolidated-invoice'
  | '386-prepayment-invoice'
  | '387-hire-invoice'
  | '388-tax-invoice'
  | '389-self-billed-invoice'
  | '390-delcredere-invoice'
  | '393-factored-invoice'
  | '394-lease-invoice'
  | '395-consignment-invoice'
  | '396-factored-credit-note'
  | '420-optical-character-reading-(ocr)-payment-credit-note'
  | '456-debit-advice'
  | '457-reversal-of-debit'
  | '458-reversal-of-credit'
  | '527-self-billed-debit-note'
  | '532-forwarders-credit-note'
  | '553-forwarders-invoice-discrepancy-report'
  | '575-insurers-invoice'
  | '623-forwarders-invoice'
  | '633-port-charges-documents'
  | '751-invoice-information-for-accounting-purposes'
  | '780-freight-invoice'
  | '817-claim-notification'
  | '870-consular-invoice'
  | '875-partial-construction-invoice'
  | '876-partial-final-construction-invoice'
  | '877-final-construction-invoice'
  | '935-customs-invoice';
